KEP-23: Enabling and disabling features in operators (#1317)
kudobuilder/kudo: keps/0023-enable-disable-features.md

---
kep-number: 23
title: Enabling and Disabling features in KUDO Operators
authors:
  - "@zmalik"
owners:
  - "@zmalik"
creation-date: 2020-01-28
last-updated: 2020-01-28
status: provisional
---

# Enabling and Disabling features in KUDO Operators

## Summary 

> KUDO enables Operator developers to create Operators and develop features for them. 

Reviewer: maybe it's rather exposing features of the underlying applications?

Author: SGTM
I just didn't want to use words like `extend operators` which would diverge from the word `feature`

The goal of this KEP is to facilitate operator developers to enable and disable the features in a deterministic way.

## Motivation

> We want to provide operator users a way to enable and disable features in their operator. And right now there is no operator-user friendly or operator-developer friendly way to do it. 

Reviewer 2: nit: _feature_  is an overloaded word. maybe we can take the definition from the L26 and put it here, making it clearer for the reader?

Author: L26 had the _example-definition_ I just added a line above `A feature in this KEP is defined as a set of Kubernetes objects.`  WDYT?

### Goals

Given a feature, `A` is composed of Kubernetes object `x`, `y` and `z`
 
- create or update kubernetes objects  `x`, `y` and `z`, when feature `A` is enabled.
- delete kubernetes objects  `x`, `y` and `z`, when feature `A` is disabled.

The goal of this KEP is to provide a way to create and delete objects `x`, `y` and `z`. 

### Non-Goals

- complex logic like loops to trigger multiple features
- enable/disable nested features
- evaluate multiple conditions to enable one feature

## Proposal

### Using Instance Parameters

In KUDO, specific plans can be triggered when updating a parameter. Plans are specifically formed by phases, steps and utimately by tasks.

> And when talking about enabling/disabling features, we are eventually talking about the `Kind` of the Task. Currently, we have `Apply`, `Delete` and `Pipe` tasks kind.

Reviewer 2: nit: there is also a `Dummy` task 😉

Enabling a feature is running a task of `Apply` kind and disabling is running a task of `Delete`

> We propose introducing a new Task kind named `Feature`. The Feature kind handles Apply or Delete resources based on the value of a parameter in the Operator Instance.

Reviewer: I would prefer the name FeatureToggle but no strong opinion here :)

Reviewer 2: I like just `Toggle` if we're going with the proposed approach

Author: sounds good to me! if I get more +1 to that I will change to `FeatureToggle`
My only reason for using `Feature` was to keep using one-word `Task` kinds. Even if underneath implementation isn't as simple as the one word might represent

Reviewer 3: I'd probably prefer `Toggle` as well, it describes what it does and it's a one-work task. `FeatureToggle` would be my second choice.

A `Feature` task will look like:

```
tasks:
  - name: mirror-maker-configuration
    kind: Feature
    spec:
      parameter: MIRROR_MAKER_ENABLED
      resources:
        - mirror-maker-cm.yaml

>   - name: mirror-maker

Reviewer: could this ultimately be one task with two resources? 🤔

Author: yes, but as we are moving towards making more sense of steps, I wanted to keep a `configuration` step and another `app` step 
but to answer your question yes, it could be only one task and one step for this example

Reviewer 2: I think, for the purpose of understanding this KEP, one task (with two resources) would be sufficient.

Reviewer 2: +1 to _one-word-kinds_. and my _heavy-plus_ goes to `kind: Toggle`, see my [reasoning](https://github.com/kudobuilder/kudo/pull/1317#pullrequestreview-350131081) for it.

    kind: Feature
    spec:
      parameter: MIRROR_MAKER_ENABLED
      resources:
        - mirror-maker.yaml
```

the parameter used in task would be defined in the parameters for the Operator, and must evaluate to a boolean value.
```
apiVersion: kudo.dev/v1beta1
parameters:
  - name: MIRROR_MAKER_ENABLED

>     default: "false"

Reviewer: as part of this we should add a linting rule that this could be only true/false? Too bad we don't have types here, because then anyone can come and change the value of this param to "whatever" and I am not sure what we do then...

Author: linting would help with default values maybe, but when somebody comes and update with a random value that cannot be directly mapped to a boolean value, we should fail the plan.

for when we have types, if we decide to do that, it would be easier as the error could be direct to parameter type related. And not a plan failing.

Reviewer 2: > change the value of this param to "whatever" and I am not sure what we do then...

that would be a `FATAL_ERROR` 🤷‍♀

Reviewer 4: @zmalik what do you think about having an implicit param for a toggle task?   why force a number of tasks on operator dev that "must" be followed in a specific pattern?

we could have an implicit param based on the task name for `mirror-maker-configuration`... which automatically follows the "rules" which also doesn't need to be linted because it is correct by convention.

I propose that `mirror-maker-configuration` or `mirror-maker-configuration-enabled` based on the task name.  To enable or disable is to pass `-p mirror-maker-configuration=false` to install/update with a default being true.

Reviewer 4: regarding **default** perhaps the default is defined in the Toggle task.  Default `enabled` | `disabled` or `Enabled: true|false`

Reviewer 3: I'd prefer it to be explicit. 
- It keeps everything you can define with `-p` in the params.yaml
- It gives the operator developer more freedom with regards to naming their parameters. (Which might not be a good thing ;)
- It allows one to use one param on multiple toggle-tasks.

If we go with implicits, there are a couple questions:
- Is this implicit variable allowed to be used somewhere else in templates?
- We probably would need validation to prevent a param with the same name to be defined in `params.yaml`

    trigger: mirror-maker-plan
```

And the plan triggered by the parameter will be the one using `Feature` tasks in their steps.

```
plans:
  mirror-maker-plan:
    strategy: serial
    phases:
      - name: deploy-mirror-maker
        strategy: serial
        steps:
          - name: configuration
            tasks:
              - mirror-maker-configuration
          - name: deploy
            tasks:
              - mirror-maker
```

> In the tasks is where we can use the `Feature` kind. Which will apply the resources rendered by templates `mirror-maker.yaml` and `mirror-maker-cm.yaml` in case the value of parameter `MIRROR_MAKER_ENABLED` is true. 

Reviewer 2: wdyt:
```suggestion
If the `spec.parameter` is `"true"`, feature task acts like `Apply` task, applies the provided resources and waits for them to become healthy. If the `spec.parameter` is `"false"` it acts like a `Delete` task, deleting the resources.
```

In case, the value is `false` it will delete the resources rendered by `mirror-maker.yaml` and `mirror-maker-cm.yaml`. 

In case the parameter value is **true**:

If there are objects to created the Task will be marked successful, once those are created.
If there are no objects to be created the Task will be marked as successful. 

In case the parameter value is **false**:

> If there are objects to be deleted the Task will be marked as successful, once those are marked for Deletion. 

Reviewer: so it basically does client.Delete for all resources specified in the resources of that task? Maybe that's a better description? WDYT?

If there are no objects to be deleted the Task will be marked as successful. 

### Limitations

However the `Feature` task will have limitations like:

- KUDO Operator must define the parameter which will be evaluated as a boolean.
- The defined parameter must evaluate to a boolean value
- Multiple parameters can trigger the same plan, but only one parameter will be evaluated

For the last use case, e.g., we can have a parameter `MIRROR_MAKER_MEM` which triggers the `mirror-maker-plan` but
the parameter which will decide in `Feature Task` if the kubernetes objects are created/updated or deleted will be `MIRROR_MAKER_ENABLED`

## Alternatives

> There are two alternatives to this approach

Reviewer 2: Here is another alternative:
The current suggestion encapsulates the `if-else` logic in a new task type. However, both branches (_true_ and _false_) are already covered by the `Apply` and `Delete` tasks so we might as well put the `if-else` logic directly into the plan like:

```
plans:
mirror-maker-plan:
  strategy: serial
  phases:
    - name: mirror-maker-phase
      strategy: serial
      steps:
        - name: apply-mirror-maker
          tasks:
            - apply-mirror-maker-configuration
            - apply-mirror-maker
          when: (MIRROR_MAKER_ENABLED == "true")
        - name: delete-mirror-maker
          tasks:
            - delete-mirror-maker-configuration
            - delete-mirror-maker
          when: (MIRROR_MAKER_ENABLED == "false")
```
Here, e.g. `apply-mirror-maker` is a simple `Apply` and `delete-mirror-maker` a `Delete` kind of task (omited here for brevity). 

**Pros:**
 - we don't need a new task type
 - more conditions flexibility: it's easy to extend the `when` to whatever we need it to be

**Cons:**
 - YAMLy programming language 🤮
 - hard to debug

P.S. I'm not advocating for it. I think it is awful  😂 But It is semantically the same and makes it clear that we're not in the "declarative world" anymore.

Reviewer 3: Good point for an alternative. And it sounds tempting, because it's kind of powerful, but as you said, hard to debug, and if we want programming, it shouldn't be in yaml.

Author: added the alternative!

> - Add feature enablement field in Instance and drop using parameters for enabling features in operators. 

Reviewer 2: I read this sentence twice but I still don't understand how it is meant. Could you please elaborate?

Author: let me re-write with some example there, but basically its to use something with clear intention to enable a feature like `Instance.spec.enabledFeatures` to get a list of enabled features and drop using `Instance.spec.parameters`

That would mean redesigning the instance updates architecture. As right now, this is the only way to updates the KUDO Instances. 
But the impact of introducing the new field would be quite limiting as it won't provide much more than what we already have with parameters 

> - Reconcile the world. KUDO creates a state of an instance based on `Instance` and `OperatorVersion` objects. And on

Reviewer 2: If that were always possible, why bother with serial phases/steps/tasks at all? If you could always define the outcome state and ignore individual steps and its order then you wouldn't need KUDO.

Author: state can include phases/steps/tasks. And reconciling a state doesn't mean to ignore the order to reach that state.

              - mirror-maker
```

If the `spec.parameter` is `"true"`, feature task acts like `Apply` task, applies the provided resources and waits for them to become healthy. If the `spec.parameter` is `"false"` it acts like a `Delete` task, deleting the resources.

> In case, the value is `false` it will delete the resources rendered by `mirror-maker.yaml` and `mirror-maker-cm.yaml`. 

Reviewer 2: wdyt:
```suggestion
So in the above example, if the parameter is `"false"` it will delete the resources rendered by `mirror-maker.yaml` and `mirror-maker-cm.yaml`. 
```
